components/constellation/constellation.rs

```diff
@@ -1376,6 +1384,94 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
         }
     }
 
+    fn handle_manual_redirect(&mut self, id: PipelineId, load_data: LoadData) {
+        let mut req_init = RequestInit {
+            url: load_data.url.clone(),
+            method: load_data.method,
+            destination: Destination::Document,
+            credentials_mode: CredentialsMode::Include,
+            use_url_credentials: true,
+            origin: load_data.url.clone(),
+            pipeline_id: Some(id),
+            referrer_url: load_data.referrer_url,
+            referrer_policy: load_data.referrer_policy,
+            headers: load_data.headers,
+            body: load_data.data,
+            redirect_mode: RedirectMode::Manual,
+            .. RequestInit::default()
+        };
+
+        let res_init = ResponseInit {
+            url: load_data.url.clone(),
+            .. ResponseInit::default()
+        };
+
+        // keep looping until we get a response that doesn't redirect
+        loop {
+            let (sender, receiver) = ipc::channel().expect("Failed to create IPC channel!");
+            if let Err(e) = self.public_resource_threads.send(
+                                CoreResourceMsg::FetchRedirect(req_init.clone(),
+                                res_init.clone(),
+                                sender)) {
+                warn!("Exit resource thread failed ({})", e);
+            }
+            match receiver.recv() {
```

Comment:
One problem here is that this is a synchronous interaction with the network thread. This means that the constellation will not be able to process anything else besides a single network request at a time, which will be very inefficient. We should be using a mechanism like script's `NetworkListener`, which receives notifications on the network thread and forwards them to the script thread's event loop via a sender. This means that the constellation will need to keep track of in-progress loads so it can know how to deal with them from the regular event loop.